File: scripts/demo_research_workflow.js
/**
 * Demonstrates research workflow through real browser interaction.
 * Uses WebdriverIO for reliable browser automation.
 */
import { remote } from 'webdriverio';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

class ResearchDemo {
    constructor() {
        this.url = 'http://localhost:7860';
        this.resultsDir = path.join(__dirname, '../demo_results');
        this.currentResultsDir = null;
    }

    async setup() {
        // Browser configuration
        const opts = {
            hostname: 'localhost',
            port: 4444,
            logLevel: 'info',
            capabilities: {
                browserName: 'chrome',
                'goog:chromeOptions': {
                    args: [
                        '--no-sandbox',
                        '--disable-gpu',
                        '--window-size=1920,1080'
                    ]
                }
            }
        };

        // Create results directory
        await fs.mkdir(this.resultsDir, { recursive: true });
        this.currentResultsDir = path.join(
            this.resultsDir,
            new Date().toISOString().replace(/[:.]/g, '-')
        );
        await fs.mkdir(this.currentResultsDir);

        // Launch browser
        this.browser = await remote(opts);
        
        // Set reasonable timeouts
        await this.browser.setTimeout({
            implicit: 5000,
            pageLoad: 10000,
            script: 5000
        });
    }

    async runDemo() {
        try {
            await this.setup();
            console.log('Starting research workflow demo...');

            // Step 1: Load interface
            await this.loadInterface();

            // Step 2: Create project
            await this.createProject();

            // Step 3: Configure LLM
            await this.configureLLM();

            // Step 4: Run initial research
            const results = await this.runInitialResearch();

            // Step 5: Run follow-up research
            const refinedResults = await this.runFollowupResearch();

            // Step 6: Export results
            await this.exportResults(results, refinedResults);

            console.log('Demo completed successfully!');

        } catch (error) {
            console.error('Demo failed:', error);
            // Save error screenshot
            await this.browser.saveScreenshot(
                path.join(this.currentResultsDir, 'error.png')
            );
            throw error;

        } finally {
            await this.cleanup();
        }
    }

    async loadInterface() {
        console.log('Loading NovaAegis interface...');
        await this.browser.url(this.url);
        
        // Wait for header
        const header = await this.browser.$('h1*=NovaAegis Research Assistant');
        await header.waitForDisplayed();
    }

    async createProject() {
        console.log('Creating research project...');

        // Go to Projects tab
        const projectsTab = await this.browser.$('button=Projects');
        await projectsTab.click();

        // Create group
        const newGroupBtn = await this.browser.$('button=New Group');
        await newGroupBtn.click();

        // Fill group details
        await this.browser.$('input[aria-label="Group Name"]')
            .setValue('LLM Research');
        await this.browser.$('textarea[aria-label="Description"]')
            .setValue('Research on LLM architectures and techniques');
        await this.browser.$('button=Create').click();

        // Create project
        const newProjectBtn = await this.browser.$('button=New Project');
        await newProjectBtn.click();

        // Fill project details
        await this.browser.$('input[aria-label="Project Name"]')
            .setValue('HF Papers Analysis');
        await this.browser.$('textarea[aria-label="Description"]')
            .setValue('Analysis of recent HuggingFace papers');
        await this.browser.$('text=LLM Research').click();
        await this.browser.$('button=Create').click();

        // Verify creation
        await this.browser.$('text=HF Papers Analysis')
            .waitForDisplayed();
    }

    async configureLLM() {
        console.log('Configuring LLM settings...');

        // Go to LLM Settings
        const settingsTab = await this.browser.$('button=LLM Settings');
        await settingsTab.click();

        // Configure model
        await this.browser.$('input[value="Local"]').click();
        await this.browser.$('select[aria-label="Local Model"]')
            .selectByVisibleText('llama2-7b');
        await this.browser.$('input[aria-label="Model Path"]')
            .setValue('models/llama2-7b');

        // Set parameters
        await this.browser.$('input[aria-label="Temperature"]')
            .setValue('0.7');
        await this.browser.$('input[aria-label="Max Tokens"]')
            .setValue('2000');

        // Save settings
        await this.browser.$('button=Save LLM Settings').click();
        await this.browser.$('text=Settings saved successfully')
            .waitForDisplayed();
    }

    async runInitialResearch() {
        console.log('Running initial research...');

        // Go to Research tab
        const researchTab = await this.browser.$('button=Research');
        await researchTab.click();

        // Enter research query
        const query = `
        Research recent HuggingFace papers about LLM architectures.
        Focus on:
        - Novel architecture approaches
        - Performance improvements
        - Training innovations
        Create a structured knowledge graph and provide insights.
        `;
        await this.browser.$('textarea[aria-label="Research Request"]')
            .setValue(query);

        // Configure research
        await this.browser.$('input[aria-label="Focus Level"]')
            .setValue('0.8');
        await this.browser.$('input[aria-label="Research Depth"]')
            .setValue('4');

        // Start research
        await this.browser.$('button=Begin Research').click();

        // Wait for completion
        await this.browser.$('#knowledge-graph svg')
            .waitForDisplayed({ timeout: 60000 });
        await this.browser.$('.results-container')
            .waitForDisplayed();

        // Capture results
        const results = {
            graphData: await this.captureGraphData(),
            insights: await this.browser.$('.results-container').getText(),
            timestamp: new Date().toISOString()
        };

        // Save screenshot
        await this.browser.saveScreenshot(
            path.join(this.currentResultsDir, 'initial_research.png')
        );

        return results;
    }

    async runFollowupResearch() {
        console.log('Running follow-up research...');

        // Enter follow-up query
        const query = `
        Based on the previous research, what are the most promising
        approaches for improving LLM training efficiency?
        Focus on practical implementations and benchmarks.
        `;
        await this.browser.$('textarea[aria-label="Research Request"]')
            .setValue(query);

        // Start research
        await this.browser.$('button=Begin Research').click();

        // Wait for completion
        await this.browser.$('#knowledge-graph svg')
            .waitForDisplayed({ timeout: 60000 });
        await this.browser.$('.results-container')
            .waitForDisplayed();

        // Capture results
        const results = {
            graphData: await this.captureGraphData(),
            insights: await this.browser.$('.results-container').getText(),
            timestamp: new Date().toISOString()
        };

        // Save screenshot
        await this.browser.saveScreenshot(
            path.join(this.currentResultsDir, 'followup_research.png')
        );

        return results;
    }

    async exportResults(initialResults, followupResults) {
        console.log('Exporting results...');

        // Save combined results
        const results = {
            initial_research: initialResults,
            followup_research: followupResults,
            metadata: {
                timestamp: new Date().toISOString(),
                project: 'HF Papers Analysis'
            }
        };

        await fs.writeFile(
            path.join(this.currentResultsDir, 'research_results.json'),
            JSON.stringify(results, null, 2)
        );

        // Export from UI
        await this.browser.$('button=Export').click();

        // Wait for download and move file
        const downloadPath = await this.browser.getDownloadPath();
        const exportedFile = path.join(downloadPath, 'exported_results.json');
        await fs.rename(
            exportedFile,
            path.join(this.currentResultsDir, 'exported_results.json')
        );
    }

    async captureGraphData() {
        return await this.browser.execute(() => {
            const graphElement = document.querySelector('#knowledge-graph');
            return {
                nodes: graphElement.__data__.nodes,
                edges: graphElement.__data__.edges
            };
        });
    }

    async cleanup() {
        if (this.browser) {
            await this.browser.deleteSession();
        }
    }
}

// Run demo
const demo = new ResearchDemo();
demo.runDemo().catch(console.error);

File: scripts/demo_research_workflow.py
#!/usr/bin/env python3
"""
Demonstrates complete research workflow through browser interaction.
Shows NovaAegis analyzing HuggingFace papers and building knowledge.
"""
import asyncio
from playwright.async_api import async_playwright
import structlog
from pathlib import Path
import json
import time
from datetime import datetime

logger = structlog.get_logger()

class ResearchDemo:
    """Demonstrates research workflow through browser."""
    
    def __init__(self):
        self.url = "http://localhost:7860"
        self.logger = logger.bind(component="research_demo")
        
        # Track demo state
        self.results_dir = Path("demo_results")
        self.results_dir.mkdir(exist_ok=True)
    
    async def run_demo(self):
        """Run complete research workflow demo."""
        try:
            async with async_playwright() as p:
                # Launch browser
                browser = await p.chromium.launch(headless=False)
                context = await browser.new_context()
                page = await context.new_page()
                
                # Step 1: Setup
                await self._setup_environment(page)
                
                # Step 2: Create project
                await self._create_project(page)
                
                # Step 3: Configure LLM
                await self._configure_llm(page)
                
                # Step 4: Initial research
                results = await self._run_initial_research(page)
                
                # Step 5: Follow-up research
                refined_results = await self._run_followup_research(page)
                
                # Step 6: Export results
                await self._export_results(page, results, refined_results)
                
                # Clean up
                await context.close()
                await browser.close()
                
        except Exception as e:
            self.logger.error("demo_failed", error=str(e))
            raise
    
    async def _setup_environment(self, page):
        """Setup research environment."""
        self.logger.info("setting_up_environment")
        
        # Load interface
        await page.goto(self.url)
        await page.wait_for_selector("h1:has-text('NovaAegis Research Assistant')")
        
        # Create results directory for this run
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.current_results_dir = self.results_dir / timestamp
        self.current_results_dir.mkdir()
    
    async def _create_project(self, page):
        """Create research project."""
        self.logger.info("creating_project")
        
        # Go to Projects tab
        await page.click("text=Projects")
        await page.click("button:has-text('New Group')")
        
        # Create LLM Research group
        await page.fill("input[aria-label='Group Name']", "LLM Research")
        await page.fill("textarea[aria-label='Description']", "Research on LLM architectures and techniques")
        await page.click("button:has-text('Create')")
        
        # Create project
        await page.click("button:has-text('New Project')")
        await page.fill("input[aria-label='Project Name']", "HF Papers Analysis")
        await page.fill("textarea[aria-label='Description']", "Analysis of recent HuggingFace papers")
        await page.click("text=LLM Research")  # Select group
        await page.click("button:has-text('Create')")
        
        # Verify creation
        await page.wait_for_selector("text=HF Papers Analysis")
    
    async def _configure_llm(self, page):
        """Configure LLM settings."""
        self.logger.info("configuring_llm")
        
        # Go to LLM Settings
        await page.click("text=LLM Settings")
        
        # Configure local model
        await page.click("text=Local")
        await page.select_option("select[aria-label='Local Model']", "llama2-7b")
        await page.fill("input[aria-label='Model Path']", "models/llama2-7b")
        
        # Set parameters
        await page.fill("input[aria-label='Temperature']", "0.7")
        await page.fill("input[aria-label='Max Tokens']", "2000")
        
        # Save settings
        await page.click("button:has-text('Save LLM Settings')")
        await page.wait_for_selector("text=Settings saved successfully")
    
    async def _run_initial_research(self, page):
        """Run initial research query."""
        self.logger.info("running_initial_research")
        
        # Go to Research tab
        await page.click("text=Research")
        
        # Enter research query
        await page.fill("textarea[aria-label='Research Request']", """
        Research recent HuggingFace papers about LLM architectures.
        Focus on:
        - Novel architecture approaches
        - Performance improvements
        - Training innovations
        Create a structured knowledge graph and provide insights.
        """)
        
        # Configure research
        await page.fill("input[aria-label='Focus Level']", "0.8")
        await page.fill("input[aria-label='Research Depth']", "4")
        
        # Start research
        await page.click("button:has-text('Begin Research')")
        
        # Wait for completion
        await page.wait_for_selector("#knowledge-graph svg", timeout=60000)
        await page.wait_for_selector(".results-container")
        
        # Capture results
        results = {
            "graph_data": await self._capture_graph_data(page),
            "insights": await page.text_content(".results-container"),
            "timestamp": datetime.now().isoformat()
        }
        
        # Save screenshot
        await page.screenshot(
            path=str(self.current_results_dir / "initial_research.png")
        )
        
        return results
    
    async def _run_followup_research(self, page):
        """Run follow-up research based on initial findings."""
        self.logger.info("running_followup_research")
        
        # Enter follow-up query
        await page.fill("textarea[aria-label='Research Request']", """
        Based on the previous research, what are the most promising
        approaches for improving LLM training efficiency?
        Focus on practical implementations and benchmarks.
        """)
        
        # Start research
        await page.click("button:has-text('Begin Research')")
        
        # Wait for completion
        await page.wait_for_selector("#knowledge-graph svg", timeout=60000)
        await page.wait_for_selector(".results-container")
        
        # Capture results
        results = {
            "graph_data": await self._capture_graph_data(page),
            "insights": await page.text_content(".results-container"),
            "timestamp": datetime.now().isoformat()
        }
        
        # Save screenshot
        await page.screenshot(
            path=str(self.current_results_dir / "followup_research.png")
        )
        
        return results
    
    async def _export_results(self, page, initial_results, followup_results):
        """Export and save research results."""
        self.logger.info("exporting_results")
        
        # Save results
        results_file = self.current_results_dir / "research_results.json"
        results_file.write_text(json.dumps({
            "initial_research": initial_results,
            "followup_research": followup_results,
            "metadata": {
                "timestamp": datetime.now().isoformat(),
                "project": "HF Papers Analysis"
            }
        }, indent=2))
        
        # Export from UI
        await page.click("button:has-text('Export')")
        
        # Wait for download
        download = await page.wait_for_event("download")
        await download.save_as(
            str(self.current_results_dir / "exported_results.json")
        )
    
    async def _capture_graph_data(self, page):
        """Capture knowledge graph data."""
        return await page.evaluate("""
            () => {
                const graphElement = document.querySelector("#knowledge-graph");
                return {
                    nodes: graphElement.__data__.nodes,
                    edges: graphElement.__data__.edges
                };
            }
        """)

async def main():
    """Run research workflow demo."""
    demo = ResearchDemo()
    await demo.run_demo()

if __name__ == "__main__":
    asyncio.run(main())